### ISSUE TITLE
kernel-install eagerly resolves symlinks (since v254)

### ISSUE THREAD

**Issue by genes1122** [systemd/systemd#29317]
### systemd version the issue has been seen with

254

### Used distribution

openSUSE Tumbleweed

### Linux kernel version used

kernel-default-6.5.4-1.1

### CPU architectures issue was seen on

x86_64

### Component

kernel-install

### Expected behaviour you didn't see

I expected the upstream provided script /usr/lib/kernel/install.d/50-dracut.install to find the pregenerated initrd file /boot/initrd-6.5.4-1-default.

### Unexpected behaviour you saw

The script looked in /usr/lib/modules/6.6.0-rc2-2.gcd00403-default and did not find the pregenerated initrd file.

### Steps to reproduce the problem

Use kernel-install to remove an entry from the efi partition, then add it back with kernel-install add  -v option.  In version 253, the pregenerated initrd is copied, in 254, it is regenerated.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Version 253:
> sudo kernel-install-old -v add 6.5.4-1-default /boot/vmlinuz-6.5.4-1-default
...
+/etc/kernel/install.d/50-dracut.install add 6.5.4-1-default /efi/2ebf7cdd6cdb4335b16f89f43f467a48/6.5.4-1-default /boot/vmlinuz-6.5.4-1-default
...

Version 254:
> sudo kernel-install -v add 6.5.4-1-default /boot/vmlinuz-6.5.4-1-default
kernel image file (/usr/lib/modules/6.5.4-1-default/vmlinuz) set via command line.
Plugin arguments: add 6.5.4-1-default /efi/2ebf7cdd6cdb4335b16f89f43f467a48/6.5.4-1-default /usr/lib/modules/6.5.4-1-default/vmlinuz

Broken script fragment from /usr/lib/kernel/install.d/50-dracut.install:

add)
    if [[ $IMAGE == "uki.efi" ]]; then
        IMAGE_PREGENERATED=${KERNEL_IMAGE%/*}/uki.efi
    else
        IMAGE_PREGENERATED=${KERNEL_IMAGE%/*}/initrd
    fi
    if [[ -f ${IMAGE_PREGENERATED} ]]; then
        # we found an initrd or uki.efi at the same place as the kernel
        # use this and don't generate a new one
        [[ $KERNEL_INSTALL_VERBOSE == 1 ]] && echo \
            "There is an ${IMAGE} image at the same place as the kernel, skipping generating a new one"
        cp --reflink=auto "$IMAGE_PREGENERATED" "$BOOT_DIR_ABS/$IMAGE" \
            && chown root:root "$BOOT_DIR_ABS/$IMAGE" \
            && chmod 0600 "$BOOT_DIR_ABS/$IMAGE" \
            && exit 0

If this change is intentional, please let the dracut team know that their script no longer works as intended.

Regards,

Gene
```

**Comment by yuwata** [comment]
Is `/boot/vmlinuz-6.5.4-1-default` a symlink to `/usr/lib/modules/6.5.4-1-default/vmlinuz` ?

**Comment by genes1122** [comment]
Yes.

# ls -l /boot/vmlinuz-6.5.4-1-default 
lrwxrwxrwx 1 root root 42 Sep 24 12:42 /boot/vmlinuz-6.5.4-1-default -> ../usr/lib/modules/6.5.4-1-default/vmlinuz

Gene

**Comment by yuwata** [comment]
The behavior is actually changed when kernel-install is rewritten by C (so since v254), now symlinks in paths are always resolved.
I'd like to keep the current behavior, to support --root and/or --image option in the future.

@keszybz @DaanDeMeyer WDYT?

**Comment by keszybz** [comment]
It seems that the this was always very fragile and `50-dracut.install` was relying on some specific aspect of behaviour.

Nevertheless, I don't think we should resolve symlinks like this. In particular, because the file *name* changes, and the user might expect to see the kernel added under the name specified, not under the final name that the symlink chain resolves to. What are we trying to achieve by resolving the symlink eagerly?

**Comment by DaanDeMeyer** [comment]
I'm not sure, I think it makes sense to resolve the symlink. It's what we generally do everywhere else these days as well. While perhaps not really applicable for the initrd, the uki.efi that the dracut install script also looks for could just as well be located in /usr/lib/modules and would only be found if the symlink was resolved.

**Comment by keszybz** [comment]
> I'm not sure, I think it makes sense to resolve the symlink.

Yeah, but why? In our internal tools we don't always do this. For most cases, there is no difference, but there are some where it matters. For example, `systemctl enable` / `link` would be broken by eager symlink resolution. For our internal code its different, because we know when we only care about the final file, but here we are interfacing with external code, which might (and apparently does) care about the original file name.

**Comment by genes1122** [comment]
> > I'm not sure, I think it makes sense to resolve the symlink.
> 
> Yeah, but why? In our internal tools we don't always do this. For most cases, there is no difference, but there are some where it matters. For example, `systemctl enable` / `link` would be broken by eager symlink resolution. For our internal code its different, because we know when we only care about the final file, but here we are interfacing with external code, which might (and apparently does) care about the original file name.

I hope it's ok for me to add my thoughts.  I agree with keszybz's observation here.  The kludge I put into my override script is to just substitute /boot for  ${KERNEL_IMAGE%/*} in the line, " IMAGE_PREGENERATED=${KERNEL_IMAGE%/*}/initrd".  That restores the original behavior.

**Comment by bluca** [comment]
So what do we do here? Punt it to the next release?

**Comment by yuwata** [comment]
After re-reading the above discussion, now I agree with @keszybz's comments.
Fix is waiting in #30111.
